### PR TITLE
Don't allocate memory when sorting a list with a comparison

### PR DESCRIPTION
**Pull request by mderoy**
Our users noticed that sorting a list with the upgraded mono runtime would cause allocations. After investigating I've found that the new classlibs are allocating by creating an IComparer wrapper object around the Comparison. We could however instead call the array sort method that accepts a comparison object. In our downstream mono 5.0 branch I added an internal Sort function to the Array class and reference that function from list...upstream it would appear things have been refactored to use the corert Array libraries which includes refactoring the array sort functions into an ArraySortHelper class. Similar to my downstream changes we can call the ArraySortHelper function that accepts a comparision object instead of allocating.

### CODE REVIEW

**Comment by akoeplinger** [comment]
whitelist

File: mcs/class/referencesource/mscorlib/system/collections/generic/list.cs [hunk 1]
```diff
@@ -997,8 +997,7 @@ public void Sort(IComparer<T> comparer)
             Contract.EndContractBlock();
 
             if( _size > 0) {
-                IComparer<T> comparer = new Array.FunctorComparer<T>(comparison);
-                Array.Sort(_items, 0, _size, comparer);
+                ArraySortHelper<T>.Sort(_items, 0, _size, comparison);
```
Comment (akoeplinger) [inline mcs/class/referencesource/mscorlib/system/collections/generic/list.cs]: could you please do this in an `#if MONO` block? it makes merging upstream referencesource changes easier :)

Comment (mderoy) [inline mcs/class/referencesource/mscorlib/system/collections/generic/list.cs]: Gladly,
Pushed that change for you :)

**Comment by mderoy** [comment]
@akoeplinger looks like something may have gone run on the test or on master?
I don't see any failures on the linux links, and the windows one has a socket error that doesn't seem related. maybe we can rerun this?

**Pull request merged**